File: origin/YuGiOhCard/View/CoinView.m
#import "CoinView.h"
#import "StringConsts.h"

@interface CoinView()
@end

@implementation CoinView

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self makeUI];
    }
    return self;
}

-(void) makeUI {
    self.backgroundColor = [UIColor clearColor];
    
    self.imgCoin = [[UIImageView alloc] initWithFrame:CGRectMake(50, 0, self.frame.size.width-100, self.frame.size.height)];
    [self.imgCoin setContentMode:UIViewContentModeScaleAspectFit];
    [self.imgCoin setUserInteractionEnabled:YES];
    UITapGestureRecognizer * singleTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(closeClicked:)];
    [self.imgCoin addGestureRecognizer:singleTap];
    
    int coinval = arc4random() % 2 + 1;
    NSString * imgName = [NSString stringWithFormat:@"coin%d", coinval];
    [self.imgCoin setImage:[UIImage imageNamed:imgName]];
    
    [self addSubview:self.imgCoin];
}

-(void) closeClicked: (id) sender {
    if ([self.delegate respondsToSelector:@selector(doneCoin)]) {
        [self.delegate doneCoin];
    }
}

@end

File: swift/YuGiOhCard/View/CoinView.swift
import UIKit

@objc
protocol CoinDelegate: NSObjectProtocol {
    optional func doneCoin()
}

class CoinView: UIView {

    var imgCoin: UIImageView?
    var delegate: CoinDelegate?

    override init(frame: CGRect) {
        super.init(frame: frame)
        self.makeUI()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.makeUI()
    }
    
    
    func makeUI() {
        self.backgroundColor = UIColor.clearColor()
        self.imgCoin = UIImageView(frame: CGRectMake(50, 0, self.frame.size.width-100, self.frame.size.height))
        self.imgCoin!.contentMode = UIViewContentMode.ScaleAspectFit
        self.imgCoin!.userInteractionEnabled = true
        let singleTap = UITapGestureRecognizer(target: self, action: "closeClicked:")
        self.imgCoin!.addGestureRecognizer(singleTap)

        let coinval = arc4random() % 2 + 1
        let imgName = "coin\(coinval)"
        self.imgCoin!.image = UIImage(named: imgName)
        self.addSubview(self.imgCoin!)
    }
    
    func closeClicked(sender: AnyObject) {
        self.delegate?.doneCoin?()
    }

}
